flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java

```diff
@@ -114,6 +114,7 @@ public void barrierAnnouncement(
 			lastSeenBarrier = barrier.getId();
 			firstBarrierArrivalTime = getArrivalTime(barrier);
 		}
+		activeController = chooseController(barrier);
```

Comment:
? 

Why is it not enough to do this in `preProcessFirstBarrierOrAnnouncement`?

(as in other places: a unit test would be helpful)

Reply:
Consider a checkpoint with two barriers from two channels: UC and AC:
1. Process AC announcement. Set `SingleCheckpointBarrierHandler.currentCheckpointId`. Don't timeout. 
2. Process UC barrier (no announcements). Because `currentCheckpointId` was set `checkSubsumedCheckpoint` does nothing, i.e. `preProcessFirstBarrierOrAnnouncement` not called and controller so `activeController` == AC
3. Further process (UC) barrier - now AC controller is processing UC barrier

> (as in other places: a unit test would be helpful)

I agree, will add it.

Reply:
But doesn't it mean we should support time outing on every UC barrier? And it looks like we are doing that:
```
	@Override
	public Optional<CheckpointBarrier> barrierReceived(InputChannelInfo channelInfo, CheckpointBarrier barrier) throws IOException, CheckpointException {
		if (barrier.getCheckpointOptions().isUnalignedCheckpoint() && activeController == alignedController) {
			barrier = barrier.asUnaligned();
			switchToUnaligned(channelInfo, barrier);
			activeController.barrierReceived(channelInfo, barrier);
			return Optional.of(barrier);
		}
```

Reply:
But there is a `preProcessFirstBarrier` before the switch in `barrierReceived`.

Reply:
I've added `AlternatingControllerTest.testSwitchToUnalignedByUpstream`.

Reply:
I still don't get this fix.
> But there is a preProcessFirstBarrier before the switch in barrierReceived.

Yes, but my intention was that in that case, `preProcessFirstBarrier` would be called on the AlignedController. Next on the first barrier, we would switch to unaligned controller, which is doing this:
```
		// alignedController might has already processed some barriers, so "migrate"/forward those calls to unalignedController.
		unalignedController.preProcessFirstBarrier(channelInfo, barrier);
		for (InputChannelInfo blockedChannel : blockedChannels) {
			unalignedController.barrierReceived(blockedChannel, barrier);
		}
```
so `preProcessFirstBarrier` would be eventually called on the unaligned controller.